ENH: add cobyqa to scipy.optimize.

ragonneau opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes #19918.

#### What does this implement/fix?
<!--Please explain your changes.-->
This PR adds the COBYQA method to the `scipy.optimize` module.

#### Additional information
<!--Any additional information you think is important.-->

The initial results of the benchmarks are posted in https://github.com/scipy/scipy/issues/19918#issuecomment-1903708120.

Commit:
ENH: add cobyqa to scipy.optimize.

See #19918.

dschmitz89 commented:
Thanks for the PR @ragonneau , this was really quick! Just a quick lunch break review regarding the API.

If you edit your message above to "Closes #19918", github will automatically link the issue. Regarding lint failures: locally, these checks can be run via `python dev.py lint`.

It looks like COBYQA uses different variable names for the tolerances than scipy's solvers. Is `tol` something similar to scipy's `ftol` or `xtol` (see [here](https://docs.scipy.org/doc/scipy/reference/optimize.minimize-powell.html#optimize-minimize-powell) for the classic Powell algorithm)?

I did not see a test where the objective uses additional user supplied arguments. Does COBYQA already support that?

ragonneau commented:
Hi @dschmitz89,

> If you edit your message above to "Closes #19918", github will automatically link the issue. Regarding lint failures: locally, these checks can be run via `python dev.py lint`.

I will check the lint results and modify COBYQA accordingly. I first focused on the primary unit tests and the benchmarks.

> It looks like COBYQA uses different variable names for the tolerances than scipy's solvers. Is `tol` something similar to scipy's `ftol` or `xtol` (see [here](https://docs.scipy.org/doc/scipy/reference/optimize.minimize-powell.html#optimize-minimize-powell) for the classic Powell algorithm)?

Yes, the option that controls the main stopping criterion tolerance is called `radius_final` in COBYQA. I did this to indicate clearly that this "tolerance" is a final trust-region radius. In `_minimize_cobyqa`, however, I renamed this options as `tol`, to match the SciPy standards. Did I misunderstand what was expected here?

Thanks,
Tom.

andyfaff commented:
In the past `tol` has taken on various meanings for various optimisers. I'd like to work towards more uniformity. Here are some tols that one could use:

- fatol, Absolute error in func(xopt) between iterations that is acceptable for convergence.
- frtol, relative change in evaluated function value
- xatol, Absolute error in xopt between iterations that is acceptable for convergence.
- xrtol, relative change in parameter value
- gatol, absolute change in gradient norm
- grtol, relative change in gradient norm

I don't think it's a good idea to use `tol` for keyword, it doesn't tell us by name what it does.

 w.r.t all the other options can you peruse the other optimisers to try and achieve convergence. e.g. there is a `radius_init` keyword, this appears to do the same as `initial_trust_radius` for [`'trust-exact'`](https://docs.scipy.org/doc/scipy/reference/optimize.minimize-trustexact.html#optimize-minimize-trustexact). (Although `trust-constr` uses `initial_tr_radius`.

Commit:
Update scipy/_lib/cobyqa

ragonneau commented:
> Regarding lint failures: locally, these checks can be run via `python dev.py lint`.

I just took care of those in the COBYQA repo and updated the submodule here.

ragonneau commented:
> In the past `tol` has taken on various meanings for various optimisers. I'd like to work towards more uniformity. Here are some tols that one could use:
> 
>     * fatol, Absolute error in func(xopt) between iterations that is acceptable for convergence.
> 
>     * frtol, relative change in evaluated function value
> 
>     * xatol, Absolute error in xopt between iterations that is acceptable for convergence.
> 
>     * xrtol, relative change in parameter value
> 
>     * gatol, absolute change in gradient norm
> 
>     * grtol, relative change in gradient norm

Great, thanks for the list.

> I don't think it's a good idea to use `tol` for keyword, it doesn't tell us by name what it does.

I totally agree, this is why I did not use it in the code of COBYQA.

> w.r.t all the other options can you peruse the other optimisers to try and achieve convergence. e.g. there is a `radius_init` keyword, this appears to do the same as `initial_trust_radius` for [`'trust-exact'`](https://docs.scipy.org/doc/scipy/reference/optimize.minimize-trustexact.html#optimize-minimize-trustexact). (Although `trust-constr` uses `initial_tr_radius`.

I agree that uniformity is preferred. However, I believe that `initial_tr_radius` is much better than `initial_trust_radius`, as the latter does not really make sense mathematically. I could definitely change `radius_init` and `tol` in `_minimize_cobyqa` to `initial_tr_radius` and `final_tr_radius` (of course the code in `_minimize.py` needs to be updated correspondingly). What do you think @andyfaff?

Cheers,
Tom.

Commit:
Improve options for COBYQA

Commit:
Improve tests of COBYQA

ragonneau commented:
Hi @andyfaff,

In the updates of this PR, I proposed `initial_tr_radius` and `final_tr_radius` as the option names for COBYQA. I updated `scipy/optimize/_minimize.py` to set the default value of `options['final_tr_radius']` to `tol`. What do you think?

I really believe that `initial_trust_radius` is not a good name, as it does not make much sense mathematically. I understand that `initial_tr_radius` has its flaws (in particular, the abbreviations used are not consistent), but I believe it is directly understandable by any engineer/researcher familiar with trust-region methods.

ragonneau commented:
Hi @dschmitz89,

> I did not see a test where the objective uses additional user supplied arguments. Does COBYQA already support that?

In the updates of this PR, I improved the `scipy/optimize/tests/test_cobyqa.py`. In particular, the argument `args` of the `minimize` function is tested. What do you think of the new tests?

dschmitz89 commented:
> Hi @dschmitz89,
> 
> > I did not see a test where the objective uses additional user supplied arguments. Does COBYQA already support that?
> 
> In the updates of this PR, I improved the `scipy/optimize/tests/test_cobyqa.py`. In particular, the argument `args` of the `minimize` function is tested. What do you think of the new tests?

The new tests looks good, thank you!

There will likely be more detailed reviews in the coming weeks still.

j-bowhay commented on scipy/optimize/tests/test_cobyqa.py:
```diff
@@ -0,0 +1,121 @@
+        )
+        solution = [(4 - np.sqrt(7)) / 3, (np.sqrt(7) - 1) / 3]
+        assert_allclose(sol.x, solution, atol=1e-4)
+        assert_(sol.success, sol.message)
```
stylistic nit but I think we now prefer  just the plain old `assert`

ragonneau replied:
92486a126ed07dba7950eedbf5db0679688cf615 should resolve this.

Commit:
Replace `assert_` with `assert`

dschmitz89 commented on scipy/optimize/_cobyqa_py.py:
```diff
@@ -0,0 +1,53 @@
+
+
+def _minimize_cobyqa(fun, x0, args=(), bounds=None, constraints=(),
+                     callback=None, disp=False, maxfev=1000, maxiter=1000,
```
I do not know COBYQA in detail but in my experience for higher dimensional problems, derivative-free solvers need often a much higher function evaluation budget. What about setting it automatically to something like 1000 * dimensionality of the problem?

dschmitz89 replied:
Also could you add tests that with a very tiny budget for `maxiter` and `maxfev` the results object returns `success=False` and that the actual number of `maxfev` stays below the provided one then still? I saw that your tests contain logic to count the number of function evaluations, thanks for that!

ragonneau replied:
> I do not know COBYQA in detail but in my experience for higher dimensional problems, derivative-free solvers need often a much higher function evaluation budget. What about setting it automatically to something like 1000 * dimensionality of the problem?

You are right. In the COBYQA package, I use `500 * n` for the default `maxfev`, and `1000 * n` for the default `maxiter`. I agree these values are much better, I just copied these default from COBYLA interface. I will update these.

ragonneau replied:
> Also could you add tests that with a very tiny budget for `maxiter` and `maxfev` the results object returns `success=False` and that the actual number of `maxfev` stays below the provided one then still? I saw that your tests contain logic to count the number of function evaluations, thanks for that!

Great idea! I will do that also.

ragonneau replied:
This should be done in e737b9c1518aed996a4798b94ce1bc67692cc780. I let you resolve this conversation if you are happy with the modifications.

dschmitz89 replied:
I am happy with the changes but do not have "resolve" permissions.

dschmitz89 commented on scipy/optimize/_cobyqa_py.py:
```diff
@@ -0,0 +1,53 @@
+
+def _minimize_cobyqa(fun, x0, args=(), bounds=None, constraints=(),
+                     callback=None, disp=False, maxfev=1000, maxiter=1000,
+                     target=-np.inf, feasibility_tol=1e-8,
```
Could you add a test for the `target` termination criterion? That one should be easy to do. I guess the other tolerances are hard to test without reusing COBYQA internals. I think we should rely on tests in the actual COBYQA repo for that, do you agree?

ragonneau replied:
Yes, I agree. I did not include all the tests in the actual COBYQA repo to SciPy because they take around 20 minutes to finish. For `feasibility_tol`, we can test that `maxcv` in the result is below this value if `success` is `True`.

ragonneau replied:
This should be done in e737b9c1518aed996a4798b94ce1bc67692cc780. Concerning the tests for `feasibility_tol`, there were already present because I test that `sol.maxcv < 1e-8`, which is the default `feasibility_tol`. I let you resolve the conversation is you are happy with the modifications.

dschmitz89 replied:
Thanks for the new test, looks good!

I had edited my first message a few minutes after it was first posted and added the first paragraph about the naming of the parameter. Your update was already on the way, so you probably missed that part (my bad actually, should have probably added it below the original comment, not above ... ). Could you still adapt that?

ragonneau replied:
Indeed, I did not see the update. I had already created a test for `target`, now renamed `f_min`. Does it look ok to you?

zaikunzhang replied:
Hi @dschmitz89 and @ragonneau,

> What is called target here is called f_min for [SHGO](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.shgo.html#scipy.optimize.shgo) and [DIRECT](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.direct.html#scipy.optimize.direct). For consistency let us stick with the same parameter name here.

Indeed, I would suggest the opposite. As a person working in optimization for more than 15 years, I have never heard anyone meant target value by f_min, which exclusively mean the minimum of f. To be frank, this is a wrong name in the current context. 

In contrast, ftarget or f_target is the standard term adopted in popular optimization packages. @ragonneau please put some references here.

Thanks.

Zaikun

ragonneau replied:
Hi @zaikunzhang,

I suppose you are referring to the following.
1. BFO (https://sites.google.com/site/bfocode) from Ph. L. Toint and M. Porcelli.
2. A. Friedlander and J. M. Martínez. Convergence and complexity in nonlinear optimization. Course dispensed at Unicamp, São Paulo, Brazil. 2017. url: https://www.ic.fcen.uba.ar/elavio/friedlandermartinez.pdf.
3. R. Tanabe. "Benchmarking feature-based algorithm selection systems for black-box numerical optimization." IEEE Trans. Evol. Comput. 26.6 (2022), pp. 1321–1335. doi: [10.1109/TEVC.2022.3169770](https://doi.org/10.1109/TEVC.2022.3169770).

Note that they referred to it as `ftarget` and not `f_target`. It is clear that `ftarget` here is not possible as it does not really comply with the Python naming standard [PEP8](https://peps.python.org/pep-0008/).

Cheers,
Tom.

Commit:
Improve default values of maxfev and maxiter

Commit:
Rename `target` to `f_min`

ragonneau commented:
Hi @andyfaff and @rgommers,

I am unsure how to solve the "Linux Meson tests" issue. I did not install the test files of COBYQA in Meson because I do not want them to be detected. Running the full suite takes a long time. How should we proceed?

Cheers,
Tom.

ilayn commented:
Are there basic but nontrivial problems for the testing that won't take too much time? Alternatively, we can create our own simple test suite by just taking a subset of yours and mark them `slow` so they only run in some of the test actions. At least the input validations should be present to catch strange API edge cases.

I will probably need to open a PR on your side to change all printing to logging since we are slowly trying to leave that historical behavior to rest. Logging is a much better and powerful system to generate output and it would be nice if we do this from the get go instead of this olden way of displaying stuff.

lucascolley commented:
> I did not install the test files of COBYQA in Meson because I do not want them to be detected. Running the full suite takes a long time. How should we proceed?

Aside from what @ilayn has said regarding the tests we want for SciPy, you can resolve the CI error by adding the paths to https://github.com/scipy/scipy/blob/1e3e8c603126f1bf2e8febd567cfd374d169283b/tools/check_installation.py#L37-L43

zaikunzhang commented:
Hi @dschmitz89 and @ragonneau,

> What is called target here is called f_min for SHGO and DIRECT. For consistency let us stick with the same parameter name here.

Indeed, I would suggest the opposite. As a person working in optimization for more than 15 years, I have never heard anyone meant target value by f_min, which exclusively mean the minimum of f. To be frank, this is a wrong name in the current context.

In contrast, ftarget or f_target is the standard term adopted in popular optimization packages. @ragonneau please put some references here.

Thanks.

Zaikun

ragonneau commented:
Hi @zaikunzhang,

> In contrast, ftarget or f_target is the standard term adopted in popular optimization packages. @ragonneau please put some references here.

I suppose you are referring to the following.
1. BFO (https://sites.google.com/site/bfocode) from Ph. L. Toint and M. Porcelli.
2. A. Friedlander and J. M. Martínez. Convergence and complexity in nonlinear optimization. Course dispensed at Unicamp, São Paulo, Brazil. 2017. url: https://www.ic.fcen.uba.ar/elavio/friedlandermartinez.pdf.
3. R. Tanabe. "Benchmarking feature-based algorithm selection systems for black-box numerical optimization." IEEE Trans. Evol. Comput. 26.6 (2022), pp. 1321–1335. doi: [10.1109/TEVC.2022.3169770](https://doi.org/10.1109/TEVC.2022.3169770).

Note that they referred to it as `ftarget` and not `f_target`. It is clear that `ftarget` here is not possible as it does not really comply with the Python naming standard [PEP8](https://peps.python.org/pep-0008/).

Cheers,
Tom.

dschmitz89 commented:
> Hi @dschmitz89 and @ragonneau,
> 
> > What is called target here is called f_min for SHGO and DIRECT. For consistency let us stick with the same parameter name here.
> 
> Indeed, I would suggest the opposite. As a person working in optimization for more than 15 years, I have never heard anyone meant target value by f_min, which exclusively mean the minimum of f. To be frank, this is a wrong name in the current context.
> 
> In contrast, ftarget or f_target is the standard term adopted in popular optimization packages. @ragonneau please put some references here.
> 
> Thanks.
> 
> Zaikun

I agree that `f_min` is not a very catchy/easy to grasp parameter name. `nlopt` for example calls this parameter `stopval`: https://nlopt.readthedocs.io/en/latest/NLopt_Reference/#stopping-criteria

We cannot change it for the existing optimizers without a deprecation cycle though. It would also be great to have this stopping criterion for all optimizers in principle.

As a library, we have to find a compromise between internal consistency and "managed evolution". In my view, we should try to unify this parameter name in one go (one release) to avoid user confusion. Open to other opinions though.

andyfaff commented:
I'd prefer to use f_target for this PR and deprecate the rest

Commit:
Add tests of cobyqa

Commit:
Rename `f_min` to `f_target`

ragonneau commented:
> I'd prefer to use f_target for this PR and deprecate the rest

Hi @andyfaff,

The most recent update of this PR renames `f_min` to `f_target`. Do you see anything else I can do right now?

Cheers,
Tom.

ragonneau commented:
> I will probably need to open a PR on your side to change all printing to logging since we are slowly trying to leave that historical behavior to rest. Logging is a much better and powerful system to generate output and it would be nice if we do this from the get go instead of this olden way of displaying stuff.

Hi @ilayn,

I understand. The `disp` option in COBYQA not only displays the returned `message` but also the objective and constraint function evaluations and information about the progress of the optimization procedure. Is this the expected behavior of the `disp` option in SciPy?

Cheers,
Tom.

dschmitz89 commented:
Could we leave the `stdout -> logging` update for a follow up PR? This PR looks in good shape to me otherwise. I will run workflows once more.

rgommers commented:
> I will run workflows once more.

That just crossed paths with a breaking change in numpy - I retriggered the jobs with build failures.

rgommers commented on doc/source/reference/optimize.minimize-cobyqa.rst:
```diff
@@ -0,0 +1,8 @@
+.. _optimize.minimize-cobyqa:
+
+minimize(method='COBYQA')
+----------------------------------------
```
minor: the `---` line length needs to match the length of the line above to keep Sphinx happy

ragonneau replied:
Ok, this is updated in the most recent commit.

Cheers,
Tom.

rgommers commented on scipy/_lib/meson.build:
```diff
@@ -178,5 +181,55 @@ py3.install_sources(
+
+py3.install_sources(
+  [
+    'cobyqa/cobyqa/tests/__init__.py',
```
I don't think we want to install the tests of the separate `cobyqa` package. We should rely only on the tests of the `scipy.optimize` module, because that's the only public API.

rgommers replied:
Related: `scipy/_lib/cobyqa` should be added to the `norecursedirs` listing in `pytest.ini` in the root of the repo.

ragonneau replied:
This is updated in the most recent commit. I remember that before including the tests of COBYQA in `meson.build`, some CI was complaining that some pure Python files (the tests of COBYQA) were not included in `meson.build`. I keep this conversation open until we solve this issue.

Cheer,
Tom.

rgommers replied:
I'll have a look at that now. I think we have to whitelist those test files so it's okay with that checker that they're not installed.

rgommers reviewed (commented):
The integration looks pretty clean. I only have a few minor comments on how that is done.

The benchmarks in CI didn't actually finish - I restarted the run. There are results posted at https://github.com/scipy/scipy/issues/19918#issuecomment-1903708120 - it'd be useful to edit the PR description and cross-link those, and maybe the main takeaway(s) for the folks who just look at this PR.

Commit:
Remove COBYQA's tests

ragonneau commented:
> The benchmarks in CI didn't actually finish - I restarted the run. There are results posted at [#19918 (comment)](https://github.com/scipy/scipy/issues/19918#issuecomment-1903708120) - it'd be useful to edit the PR description and cross-link those, and maybe the main takeaway(s) for the folks who just look at this PR.

I put the link you provided in the description of this PR.

Cheers,
Tom.

Commit:
Improve description of COBYQA

Commit:
Update COBYQA to v1.1.1

ragonneau commented:
Hi @rgommers

The most recent update of this PR should patch the issue occurring in the tests.

One CI will, however, not pass since the test files of the [cobyqa](https://github.com/cobyqa/cobyqa) repo are not included in the `meson.build` files. How should we handle this?

Cheers,
Tom.

rgommers commented:
The commit right before the last one had failing CI jobs from `TestOptimizeSimple.test_nan_values[cobyqa]`. I restarted CI now.

@ragonneau if you want to submit a trivial PR with some minor rewording of docs or a typo fix (doesn't matter which one), then we can merge that PR straight away, after which new pushes to this PR should no longer need maintainer approval for CI to run.

rgommers committed:
Merge branch 'main' into ragonneau/main

rgommers committed:
TST: make checker that verifies all test files are installed happy

To verify locally, run:
```
python tools/check_installation.py build-install
```
(assuming you've built with `python dev.py build`)

[skip circle] [skip cirrus]

rgommers commented:
> One CI will, however, not pass since the test files of the [cobyqa](https://github.com/cobyqa/cobyqa) repo are not included in the `meson.build` files. How should we handle this?

The commit I just pushed fixed the problem.

ragonneau commented:
Hi @rgommers,

> The commit I just pushed fixed the problem.

Great, thank you very much! There is one more set of tests:
```
cobyqa/cobyqa/utils/tests/test_exceptions.py
cobyqa/cobyqa/utils/tests/test_math.py
```

The CI should succeed after this modification. Should I let you include them in `check_installation.py`?

Cheers,
Tom.

Commit:
Exclude all tests of COBYQA in `check_installation.py`

Commit:
Correct check_installation.py

ragonneau commented:
Hi @rgommers,

Could you look at the most recent version of `check_installation.py`? I am not sure why the CI fails, as `_lib/cobyqa/cobyqa/tests/test_models.py` is included in the blacklisted files...

Thanks for your help,
Cheers,
Tom.

rgommers committed:
MAINT: improve `tools/check_installation.py` and resolve cobyqa errors

[skip cirrus]

rgommers commented:
Sorry about that - should be fixed now. I improved the error messages while I was at it, so in case it fails again with a future submodule update, it's clearer how to resolve it.

ragonneau commented:
> Sorry about that - should be fixed now. I improved the error messages while I was at it, so in case it fails again with a future submodule update, it's clearer how to resolve it.

No trouble at all. I simply thought the `_lib/` prefix was necessary, as it seems to be employed by `array_api_compat`. Let's see the outcomes of the CI now. Thank you very much for the help!

Cheers,
Tom.

j-bowhay commented on scipy/optimize/_cobyqa_py.py:
```diff
@@ -0,0 +1,53 @@
+
+from ._optimize import _check_unknown_options
+
+__all__ = []
```
I think you can probably just remove this line to make mypy happy
```suggestion
```

ragonneau replied:
Thank you very much for your suggestion, I was not sure how to deal with mypy. I just made a new commit to also remove the empty line above and avoid linting issues.

Cheers,
Tom.

Commit:
Remove empty `__all__` variable

ragonneau commented:
Hi @andyfaff @dschmitz89,

Do you see anything else I could improve for this PR?

Cheers,
Tom.

dschmitz89 commented:
@ragonneau  I'll give this one two more weeks for comments/objections, then I will merge. Currently, the NumPy 2.0 release is drawing a lot of attention from many maintainers, so less time for regular reviews.

andyfaff commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -1188,15 +1188,15 @@ def dfunc(z):
                 jac = dfunc
 
-            sol1 = optimize.minimize(func, [1, 1], jac=jac, tol=1e-10,
+            sol1 = optimize.minimize(func, [2, 2], jac=jac, tol=1e-10,
```
I think there are several locations in this file where cobyqa needs to be added.

- Please add a `test_cobyqa` method to `CheckOptimizeParameterized`. Examine the other methods to see how they're setup.
- https://github.com/scipy/scipy/blob/main/scipy/optimize/tests/test_optimize.py#L1647
- https://github.com/scipy/scipy/blob/main/scipy/optimize/tests/test_optimize.py#L1318

andyfaff commented on scipy/optimize/_cobyqa_py.py:
```diff
@@ -0,0 +1,51 @@
+        terminated when the objective function value of a nearly feasible point
+        is less than or equal to this target.
+    feasibility_tol : float
+        Tolerance for the constraint violation.
```
Absolute tolerance or relative tolerance?

ragonneau replied:
I will update the doc. This is an absolute tolerance on the $\ell_{\infty}$-constraint violation.

andyfaff commented on scipy/optimize/_cobyqa_py.py:
```diff
@@ -0,0 +1,51 @@
+        Maximum number of function evaluations.
+    maxiter : int
+        Maximum number of iterations.
+    f_target : float
```
I assume this is an absolute tolerance?

ragonneau replied:
Yes, I will also update the doc of `f_target` to make this clear.

andyfaff replied:
Would `fa_target` be better?

andyfaff reviewed (changes requested):
I've spent an hour or two on this. The code is probably in a good state, but there are some important things that need to be addressed before merge.

ragonneau commented:
> I've spent an hour or two on this. The code is probably in a good state, but there are some important things that need to be addressed before merge.

Thank you very much for taking the time. I will respond to each of your comments.

Cheers,
Tom.

ragonneau commented on scipy/optimize/_cobyqa_py.py:
```diff
@@ -0,0 +1,51 @@
+        Final trust-region radius. It should indicate the accuracy required in
+        the final values of the variables.
+    scale : bool
+        Whether to scale the variables according to the bounds.
```
It scales the variables to lie in $[-1, 1]$ and is done only if *all* the lower and upper bounds are finite. This is useful in practice when solving badly scaled problems. For example, if you are solving a $2$-dimensional problem whose first variable lies in $[10^{-6}, 10^{-3}]$ and whose second variable lies in $[10^3, 10^6]$, it is generally a good idea to do such a scaling. This is disabled by default.

There are no issues if a variable approaches the lower bound. The scaling is internal to the COBYQA method, and the "descaled" `x` is returned to the user. Should I improve the documentation of this option?

ragonneau replied:
Hi @andyfaff,

Do you still require changes after my explanation?

andyfaff replied:
I'd like this to be explained in the docstring. A user will need to know what the scaling is for, at the moment the docstring doesn't do that. Is the scaling making sure that a gradient is calculated correctly?

ragonneau replied:
I updated the documentation of `_minimize_cobyqa` accordingly. Thank you very much.

ragonneau replied:
Also, remember that COBYQA is a derivative-free method: we do not evaluate any derivative (gradient, hessian, etc). The `scale` option simply scales the problem before starting the optimization procedure. Hence, in exact arithmetic, the results should be the same (but of course, we are not in exact arithmetic).

ragonneau commented on scipy/optimize/_cobyqa_py.py:
```diff
@@ -0,0 +1,51 @@
+        the final values of the variables.
+    scale : bool
+        Whether to scale the variables according to the bounds.
+    """
```
I will add this in the next commit. Which (future) version of SciPy are you thinking of?

lucascolley replied:
Would be `1.14.0`

ragonneau replied:
Great, thanks @lucascolley!

ragonneau commented on scipy/optimize/_minimize.py:
```diff
@@ -335,6 +340,13 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
+    algorithm is a derivative-free trust-region SQP method based on quadratic
+    approximations to the objective function and each nonlinear constraint. The
+    bounds are treated as unrelaxable constraints, in the sense that the
+    algorithm always respects them throughout the optimization process.
```
Absolutely, I will add it to the test you proposed below.

ragonneau commented on scipy/optimize/_minimize.py:
```diff
@@ -466,6 +478,10 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
         implementation of an algorithm for large-scale equality constrained
         optimization. SIAM Journal on Optimization 8.3: 682-706.
+    .. [18] Ragonneau, T. M. *Model-Based Derivative-Free Optimization Methods
+       and Software*. PhD thesis, Department of Applied Mathematics, The Hong
```
We are preparing a manuscript introducing COBYQA, but it is not yet ready. The thesis describes the algorithm in detail.

Should I leave it for the time being, and propose a PR updating the documentation when the paper is ready?

ragonneau replied:
The next commit will correct the identing issue.

ragonneau commented on scipy/optimize/_shgo.py:
```diff
@@ -273,8 +273,9 @@ def shgo(
     the ``SLSQP`` method is used. In general, it is recommended to use the
-    ``SLSQP`` or ``COBYLA`` local minimization if inequality constraints
-    are defined for the problem since the other methods do not use constraints.
+    ``SLSQP``, ``COBYLA``, or ``COBYQA`` local minimization if inequality
```
I am not sure either, but I do not think this PR is the place to correct this if this is a typo.

ragonneau commented on scipy/optimize/tests/test_cobyqa.py:
```diff
@@ -0,0 +1,168 @@
+        assert sol.fun < self.fun(solution) + 1e-3, sol
+
+        # Case where the bounds are active at the solution.
+        bounds = Bounds([5.0, 0.6], [5.5, 0.65])
```
The next commit will modify `test_minimize_bounds` to check whether each function evaluation and the returned point are inside the bounds.

ragonneau commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -213,19 +213,19 @@ def test_bfgs(self):
                         [[0, -5.25060743e-01, 4.87748473e-01],
                          [0, -5.24885582e-01, 4.87530347e-01]],
                         atol=1e-14, rtol=1e-7)
-    
```
I apologize for this; my IDE automatically deletes trailing blank characters in the files I edit... I did not see at first how many such characters it deleted.

ragonneau commented on scipy/optimize/_minimize.py:
```diff
@@ -38,8 +39,9 @@
-                    'dogleg', 'trust-ncg', 'trust-exact', 'trust-krylov']
+                    'l-bfgs-b', 'tnc', 'cobyla', 'cobyqa', 'slsqp',
+                    'trust-constr', 'dogleg', 'trust-ncg', 'trust-exact',
+                    'trust-krylov']
```
I will add it to the list.

The COBYQA method accepts the two signatures mentioned in the doc of `minimize`. This is also mentioned in the doc of the COBYQA package (see https://www.cobyqa.com/stable/ref/generated/cobyqa.minimize.html).

ragonneau commented on scipy/optimize/_cobyqa_py.py:
```diff
@@ -0,0 +1,51 @@
+
+
+def _minimize_cobyqa(fun, x0, args=(), bounds=None, constraints=(),
+                     callback=None, disp=False, maxfev=None, maxiter=None,
```
The test `test_callback_stopiteration` will pass on the next commit.

I had to modify the test slightly for the following reason: methods in derivative-free optimization usually don't perform one and only one function evaluation at each iteration. They
1. usually perform function evaluations before the beginning of the iterations to initialize models (for model-based methods), and
2. might perform zero, one, two (or more) function evaluations at each iteration.

I modified the test to pass `options={'maxfev': maxiter}` to COBYQA, and to assert that `res.nfev == ref.nfev == maxiter`. Do you agree?

ragonneau replied:
I also added an extra test in `TestCOBYQA` to check that the new syntax is also understood.

ragonneau commented on scipy/optimize/_optimize.py:
```diff
@@ -86,7 +86,7 @@ def derivative(self, x, *args):
 def _wrap_callback(callback, method=None):
     """Wrap a user-provided callback so that attributes can be attached."""
-    if callback is None or method in {'tnc', 'slsqp', 'cobyla'}:
+    if callback is None or method in {'tnc', 'slsqp', 'cobyla', 'cobyqa'}:
```
Ok I misunderstood the purpose of this code. The next commit of this PR will remove `'cobyqa'` from this list.

ragonneau replied:
Hi @andyfaff,

Do you still require changes here?

ragonneau commented on scipy/optimize/_minimize.py:
```diff
@@ -624,6 +641,8 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
         if meth in ('cobyla', '_custom'):
             options.setdefault('tol', tol)
+        if meth == 'cobyqa':
+            options.setdefault('final_tr_radius', tol)
```
I copied the behavior of COBYQA and its `tol` parameter in that case. If the user supplies both the `tol` keyword and the `final_tr_radius` options, I disregard the `tol` keyword and consider only the `final_tr_radius` option.

As far as I understand, this is the expected behavior of SciPy. Did I misunderstand something?

andyfaff replied:
I think there should be documentation in _minimize_cobyqa to this effect, i.e. final_tr_radius overwrites and trumps tol. This code will attract hundreds of users and if a small percentage don't understand options, etc, then this gives rise to a lot of issues being created.

ragonneau replied:
I updated the documentation of `_minimize_cobyqa` accordingly. Thank you very much.

ragonneau commented on scipy/optimize/tests/test_cobyqa.py:
```diff
@@ -0,0 +1,168 @@
+)
+
+
+class TestCOBYQA:
```
It seems that `'cobyqa'` is already included in `test_minimize_callback_copies_array`, as it is an element of `MINIMIZE_METHODS`. I will add COBYQA to the two other tests.

ragonneau commented on scipy/optimize/tests/test_cobyqa.py:
```diff
@@ -0,0 +1,168 @@
+        assert sol.fun < self.fun(solution) + 1e-3, sol
+        assert sol.nfev == callback.n_calls, \
+            "Callback is not called exactly once for every function eval."
+
```
I agree; I will add these tests.

Commit:
Make changes requested by @andyfaff

ragonneau commented:
I don't have the right to restart CircleCI's jobs. They should now pass, as `jupyterlite-sphinx` 0.13.1 is now released. Can someone restart them to confirm that this PR is in good shape?

Thanks,
Tom.

andyfaff commented on scipy/optimize/tests/test_cobyqa.py:
```diff
@@ -0,0 +1,246 @@
+        assert sol.maxcv == sol_new.maxcv
+        assert sol.nfev == sol_new.nfev
+        assert sol.nit == sol_new.nit
+        assert sol_new.nfev == callback_new_syntax.n_calls, \
```
I would expect the callbacks to be called the same number of times as the number of iterations (sol.nit), not the same as the number of function evaluations. The callback should get triggered at the end of an iteration loop.

andyfaff replied:
I assume that the number of iterations, `sol.nit`, is the number of times an outer loop runs? Normally the callback is called at the end of that outer loop, so the callback is called `nit` times.

> In particular for COBYQA, in one iteration, one may perform zero, one, or two function evaluations. Hence, I do not think that triggering the callback at the end of each iteration makes much sense for derivative-free methods, as no function evaluation might be performed.

I don't understand that logic. There are several derivative free minimisers that have varying numbers of function evaluations on each iteration, the callback gets triggered at the end of those iterations.

Commit:
Improve the documentation of `_minimize_cobyqa`

lucascolley committed:
Merge branch 'main' into main

lucascolley commented:
> Any idea how to resolve the issue?

I've sorted it 👍

ragonneau commented:
> > Any idea how to resolve the issue?
> 
> 
> 
> I've sorted it 👍 

Thanks @lucascolley !

andyfaff commented on scipy/optimize/_cobyqa_py.py:
```diff
@@ -0,0 +1,57 @@
+        to be within the range :math:`[-1, 1]`. If any of the lower or upper
+        bounds is infinite, the variables are not scaled.
+    """
+    from .._lib.cobyqa import minimize  # import here to avoid circular imports
```
It's not clear how this avoids circular imports.

ragonneau replied:
- The `cobyqa` module imports `scipy.optimize.lsq_linear`.
- The `scipy.optimize` submodule imports `scipy.optimize.__lsq`.
- The `scipy.optimize.__lsq` submodule imports `scipy.optimize.__lsq.lsq_linear`
- The `scipy.optimize.__lsq.lsq_linear` imports `scipy.optimize`.

I think the circular import I observed comes from there (though I am not 100% sure).

andyfaff commented:
Note, I've not reviewed the `cobyqa` submodule itself!

zaikunzhang commented:
Just curious, are we moving forward? Thank you.

andyfaff commented:
I'm happy with the additions of this PR. What I've not audited is the additions from the cobyqa git submodule. Without meaning disrespect the XZ event (https://www.akamai.com/blog/security-research/critical-linux-backdoor-xz-utils-discovered-what-to-know) demonstrates the problems with code additions that have not been vetted.
I'd just like someone to go through cobyqa and check for saneness.

dschmitz89 commented:
> I'm happy with the additions of this PR. What I've not audited is the additions from the cobyqa git submodule. Without meaning disrespect the XZ event (https://www.akamai.com/blog/security-research/critical-linux-backdoor-xz-utils-discovered-what-to-know) demonstrates the problems with code additions that have not been vetted. I'd just like someone to go through cobyqa and check for saneness.

While I am not a cybersecurity expert, I am confident to state that I do not see any backdoors in the `cobyqa` [codebase](https://github.com/cobyqa/cobyqa). Also, the identities of the contributors in that repository are known (yeah, Github accounts get hacked too ..). Since we do not rely on uploaded binaries but on the code itself (unlike the xz CVE), from my point of view COBYQA can be assessed as safe.

dschmitz89 commented:
It would be great to add this in 1.14. If all your concerns are resolved, could you approve @andyfaff ?

zaikunzhang commented:
> I'm happy with the additions of this PR. What I've not audited is the additions from the cobyqa git submodule. Without meaning disrespect the XZ event ([akamai.com/blog/security-research/critical-linux-backdoor-xz-utils-discovered-what-to-know](https://www.akamai.com/blog/security-research/critical-linux-backdoor-xz-utils-discovered-what-to-know)) demonstrates the problems with code additions that have not been vetted. I'd just like someone to go through cobyqa and check for saneness.

A probably stupid question out of curiosity: Do there exist widely accepted tools that can scan the sanity of codebases automatically? If yes, such a scan should be included in the CI/CD of libraries as important as SciPy. I would also deploy it in the daily testing of [PRIMA](http://libprima.net) if that is possible. 

Would the current AI tools help in some way?

lucascolley commented:
It sounds like everyone is happy here, with the only remaining concern from Andrew being the safety of `cobyqa`. I agree with Daniel that I think it can be deemed safe. I would be happy to press the green button, if there are no objections.

lucascolley committed:
Merge branch 'main' into ragonneau/main

dschmitz89 commented:
Allright folks, I pressed the green button now. Many thanks to @ragonneau , @zaikunzhang and the reviewers. COBYQA will be noted as a highlight of the next release!

zaikunzhang commented:
Hi @dschmitz89 and all, 

Many thanks for making this come true! 

Thank the SciPy community very much, particularly the reviewers. This is the first time we contribute a package to a prestigious open-source library like SciPy. This could not have happened without support from the community.

Thanks and congratulations to Tom @ragonneau. It has been extremely rewarding and beneficial to work on COBYQA as a coauthor/supervisor during Tom's thesis. I have learned a lot. This is a project that could not have been finished without a coauthor who has a perfect background in both computer science and applied mathematics. COBYQA took us more than two years of intensive work, from the mathematical design of the algorithm to the implementation of the package. It could not have been finished without Tom's diligence and persistence. Credits should go to Tom, and blames (if any, if ever) should come to me. 

Many thanks to all!

Zaikun Zhang 
Department of Applied Mathematics 
The Hong Kong Polytechnic University

zaikunzhang commented:
> COBYQA will be noted as a highlight of the next release!

Fantastic! 

May I suggest mentioning that COBYQA is developed by us at **the Department of Applied Mathematics, The Hong Kong Polytechnic University** in the release note? Without this, my university and funding bodies may not be very happy. 

Tom and I have had a difficult time and considerable pressure at the university while working on COBYQA (and [PRIMA](http://libprima.net) in the past four years -- four years of full-time hard work has led to no publication (yet). We have been sleeping in the past four years from the university's perspective.

It will ne also very helpful if a link could be made to https://cobyqa.com, the homepage of COBYQA.

dschmitz89 commented:
> > COBYQA will be noted as a highlight of the next release!
> 
> Fantastic! 
> 
> May I suggest mentioning that COBYQA is developed by us at **the Department of Applied Mathematics, The Hong Kong Polytechnic University** in the release note? Without this, my university and funding agencies may not be very happy. 
> 
> Tom and I have had a difficult time and considerable pressure at the university while working on COBYQA, [PRIMA](http://libprima.net), and [PDFO](https://pdfo.net) in the past four years. Four years of full-time hard work has led to no publication (yet). This means that our output in the past four years is zero in the evaluation system of the university (or any research university). From the university's perspective, we have been sleeping in the past four years with zero productivity. Mentioning the department and university in the release note will make our life a bit easier --- mainly mine, as Tom has gone to industry.
> 
> It will also be very helpful if a link could be made to https://www.cobyqa.com, the homepage of COBYQA.

Adding the URL to the docstring can easily be done. Would you open a PR? About mentioning the department and university, I would defer to @rgommers . Publications are easily citable, but institutions that produce sth for SciPy as a byproduct, I am not sure. I understand your struggle though, it is a shame how little recognition OSS gets from funding agencies.

lucascolley commented:
I think a release note such as "A new method, `cobyqa`, has been added to `scipy.optimize.minimize` - this is an interface for COBYQA, a brand new derivative-free optimization solver, designed to supersede COBYLA, from the Department of Applied Mathematics, The Hong Kong Polytechnic University." sounds reasonable. The credit is attributed as an information point about the solver, rather than for the SciPy interface. But I'm not sure whether we have any rules written down about this that I haven't read.

ilayn commented:
We should be able to squeeze the affiliation information in some docstring somewhere. The link is probably better to be included in the module docstring if we already have it. Maybe in the license or the readme of the code is also suitable.

I'd skip the *brand new* or *modern* words because they look funny after 5 years, don't ask me how I know or which F77 code I am talking about :sweat_smile:

ilayn commented:
Oh and thank you for your efforts @ragonneau and @zaikunzhang . Much appreciated.

rgommers commented:
All the above suggestions sound good to me. I see it's already been added to https://github.com/scipy/scipy/wiki/*Release-Note-Entries-for-SciPy-1.14.0-(asterisk-makes-this-appear-at-the-top)#scipyoptimize-improvements, so all good here aside from adding the https://www.cobyqa.com/ reference to the docstring.

Thanks all - and great to see this merged!

zaikunzhang commented:
Many thanks to all!

@rgommers "COBYQA" is an acronym of 
"
Constrained Optimization BY Quadratic Approximations
"
rather than 
"
COnstrained Optimization BY Quadratic Approximations
"

lucascolley commented:
> "COBYQA" is an acronym of ...

Release notes updated accordingly 👍

mdhaber commented:
Thanks everyone! One more follow-up item. I see that this PR adjusted `test_all_nograd_minimizers`, and recently since I've seen that test fail because it takes too long (20s). `pytest-fail-slow` currently enforced a 5s limit on all tests not marked `xslow` - can you adjust the test or mark it `xslow`?

ragonneau commented:
Hi @lucascolley,

I noted on the dev website of SciPy that the name of Zaikun Zhang (@zaikunzhang, the co-author of COBYQA) does not appear. Would it be possible to add it to https://github.com/scipy/scipy/blob/main/doc/source/release/1.14.0-notes.rst next to mine?

Also, it is possible to add a the link to the documentation of COBYQA (https://www.cobyqa.com) to the docstring of the method in SciPy? We discussed about it in the related issue.

Thank you very much for your kind help,
Best regards,
Tom.

lucascolley commented:
> I noted on the dev website of SciPy that the name of Zaikun Zhang ... does not appear. Would it be possible to add it to [main/doc/source/release/1.14.0-notes.rst](https://github.com/scipy/scipy/blob/main/doc/source/release/1.14.0-notes.rst?rgh-link-date=2024-05-30T20%3A44%3A12Z) next to mine?

@tylerjereddy I assume you are planning to re-run the authors script for RC2? Will it need to be manually added after the script is run, or does it make sense to include separately to the generated list, but still under the same section?

> Also, it is possible to add a the link to the documentation of COBYQA ([cobyqa.com](https://www.cobyqa.com/)) to the docstring of the method in SciPy? We discussed about it in the related issue.

Sure, would you be happy to make a PR?

andyfaff commented:
I just made the PR, #20841. @ragonneau, the authors list is automatically generated from the commit log, so any edits would have to be done manually.

rgommers commented:
> the authors list is automatically generated from the commit log, so any edits would have to be done manually.

How about we backport gh-20841 to the 1.14.x branch and add @zaikunzhang on the commit with `Co-Authored By:`? Then on the next run the tooling will do the right thing.

andyfaff commented:
How does one add authors to the commit?

rgommers commented:
`Co-Authored By:`https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

ragonneau commented:
> I just made the PR, #20841. @ragonneau, the authors list is automatically generated from the commit log, so any edits would have to be done manually.

Hi @andyfaff and @lucascolley,

Sorry I was not reactive enough, thank you for handling this!

Cheers,
Tom.
